ext/avromatic/encoder.cpp

```diff
@@ -0,0 +1,267 @@
+
+        VALUE rb_nested_value = rb_hash_aref(rb_value, rb_key);
+        ruby_value_to_datum(rb_nested_value, rb_avromatic_item_type, mapValue.value()[i].second);
+      }
+      break;
+    }
+    case avro::AVRO_FIXED:
+    {
+      const char* raw_bytes = StringValuePtr(rb_value);
+      const size_t num_bytes = RSTRING_LEN(rb_value);
+      avro::GenericFixed& fixedValue = datum.value<avro::GenericFixed>();
+      fixedValue.value().assign(raw_bytes, raw_bytes + num_bytes);
+      break;
+    }
+    default:
+      rb_raise(rb_eRuntimeError, "Unexpected field type %s", avro::toString(datum.type()).c_str());
+  }
+}
+
+void ruby_model_to_datum(VALUE rb_model, avro::GenericDatum& datum, bool is_value_schema) {
+  avro::GenericRecord& recordDatum = datum.value<avro::GenericRecord>();
+  VALUE rb_attributes = rb_funcall(rb_model, rb_intern("_attributes"), 0);
+  VALUE rb_attribute_definitions = rb_funcall(rb_model, rb_intern("attribute_definitions"), 0);
+  VALUE rb_field_names = is_value_schema ?
+    rb_funcall(rb_model, rb_intern("value_avro_field_names"), 0) :
+    rb_funcall(rb_model, rb_intern("key_avro_field_names"), 0);
+  const long num_fields = rb_array_len(rb_field_names);
+  for (long i=0; i<num_fields; ++i) {
+    VALUE rb_field_name = rb_ary_entry(rb_field_names, i);
+    VALUE rb_field_value = rb_hash_aref(rb_attributes, rb_field_name);
+    VALUE rb_attribute_definition = rb_hash_aref(rb_attribute_definitions, rb_field_name);
+    VALUE rb_avromatic_type = rb_funcall(rb_attribute_definition, rb_intern("type"), 0);
+    const int field_index = (int)recordDatum.fieldIndex(rb_id2name(SYM2ID(rb_field_name)));
+    ruby_value_to_datum(rb_field_value, rb_avromatic_type, recordDatum.fieldAt(field_index));
+  }
+}
+
+VALUE encode_model(VALUE self, VALUE rb_model, VALUE rb_is_value_schema) {
+  bool is_value_schema = rb_is_value_schema == Qtrue;
+  VALUE rb_avro_schema = rb_funcall(rb_model, is_value_schema ? rb_intern("value_avro_schema") : rb_intern("key_avro_schema"), 0);
+  const avro::ValidSchema* schema = get_cached_schema(rb_avro_schema);
+
+  avro::GenericDatum datum(schema->root());
+  try {
+    ruby_model_to_datum(rb_model, datum, is_value_schema);
+  } catch (MissingRequiredAttributeException& e) {
+    // Trigger a full validation to get all error messages
+    rb_funcall(rb_model, rb_intern("avro_validate!"), 0);
+    // The line above should raise an exception but just in case...
+    rb_raise(rb_avromatic_validation_error_class, "%s", e.what());
+  }
+
+  std::unique_ptr<avro::OutputStream> outputStream = avro::memoryOutputStream();
+  avro::EncoderPtr encoder = avro::validatingEncoder(*schema, avro::binaryEncoder());
+  encoder->init(*outputStream);
+  try {
+    avro::encode(*encoder, datum);
+  } catch (const avro::Exception &e) {
+    // This should never happen since the model should be valid at this point but just in case
+    rb_raise(rb_avromatic_validation_error_class, "%s", e.what());
```

Comment:
IIUC we probably want to return a nil here, rather than trying to execute line 253.